Do not unnecessarily break sharing during renaming

tautschnig opened the pull request:
Not all subexpressions (or types) change during renaming. Only actually write to
the result when there was a change. This preserves sharing. On SV-COMP's
ReachSafety-ECA benchmarks the time spent in goto_symex_state::rename(exprt &,
...) is now 1944 seconds, when previously it was 2326 seconds. On the other
hand, the time spent in merge_ireps did not improve, the number of recursive
operator== calls in fact was higher (6930718243 rising to from 5710509892).

The time spent in goto_symex_statet::rename(typet &, ...) increases from 530
seconds to 568 seconds, which suggests there is room for improvement.

~I will keep this PR in Draft status to make sure #3996 won't need a rebase.~

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier commented:
I think for review it would be nice to split the PR in three commits: on that makes rename_address return an optional, one for rename_type and one for rename_expr.

tautschnig committed:
Do not unnecessarily break sharing during SSA renaming (part 1: types)

Instead of editing the object in place, return an optional<typet> that
is set if, and only if, a modification took place.

tautschnig committed:
Do not unnecessarily break sharing during SSA renaming (part 2: addresses)

Instead of editing the object in place, return an optional<exprt> that
is set if, and only if, a modification took place.

tautschnig committed:
Do not unnecessarily break sharing during renaming (part 3: expressions)

Not all subexpressions (or types) change during renaming. Only actually
write to the result when there was a change. This preserves sharing.

Performance results for this and parts 1 and 2 of the renaming combined:
On SV-COMP's ReachSafety-ECA benchmarks the time spent in
goto_symex_state::rename(exprt &, ...) is now 1944 seconds, when
previously it was 2326 seconds. On the other hand, the time spent in
merge_ireps did not improve, the number of recursive operator== calls in
fact was higher (6930718243 rising to from 5710509892).

The time spent in goto_symex_statet::rename(typet &, ...) increases from
530 seconds to 568 seconds, which suggests there is room for
improvement.

tautschnig commented:
Split into three commits done and rebased onto current develop - releasing for review.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7cf48757d1e65c2e8a9e021b18fc2b33a1d040b9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102536309

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
         auto &array_type = to_array_type(component.type());
-        array_type.size() = rename<level>(std::move(array_type.size()), ns);
+        auto renamed_expr = rename<level>(array_type.size(), ns);
+        if(renamed_expr != array_type.size())
```
Is this redundant? If there were no change, wouldn't it return Empty?

tautschnig replied:
That's modified in a later commit, I just did types-only in the first commit.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+      result = std::move(pointer_type);
+    }
+  }
+  else if(type.id() == ID_symbol_type)
```
@kroening believes this is now 100% gone I think

tautschnig replied:
I guess #4264 better get in before this one then there is no doubt at all. I guess it's fairly close to the top of @kroening's list.

tautschnig replied:
Cleaned up.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 5ef434d62917fc651ef2da78af1de251c42a1671).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102537067
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+  }
+  else if(type.id() == ID_struct_tag || type.id() == ID_union_tag)
+  {
+    const symbolt &symbol = ns.lookup(to_tag_type(type));
```
lookup(tag).type === follow(tag)

tautschnig replied:
Done.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -368,9 +368,15 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
   else if(expr.id()==ID_address_of)
```
(location irrelevant) two different commit messages declare themselves "part 1"

tautschnig replied:
Fixed.

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
+    optionalt<typet> renamed_type;
+    if(
+      renamed_struct_op.has_value() &&
+      renamed_struct_op->type().id() != ID_symbol_type &&
```
Needed anymore?

tautschnig replied:
Gone.

smowton reviewed (commented):
Looks good -- my comments about redundant quality tests may in fact be subsumed by later commits.

Question: does this render the rename-required pre-pass redundant? Should we combine the two methods into one walk at this point?

Pattern question: mild preference for something that takes an `exprt &` as at present, carefully turns that into a `const exprt &`, uses that to determine if anything needs doing, and then writes to it if appropriate, rather than having the `if(ret) val = *ret;` pattern at every callsite

tautschnig commented:
I'm taking the liberty to mark this dependent-on-#4264.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -680,11 +714,27 @@ void goto_symex_statet::rename(
   // is no need to expand the struct/union tags in the type
   if(!requires_renaming(type, ns))
-    return; // no action
+    return {};
```
Is the call to `requires_renaming` still needed now? I would think the false case would correspond to the case where we get an empty optional below.

tautschnig replied:
I have added a commit to remove it, and also ported the test from #4233.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
 
   if(type.id()==ID_array)
   {
     auto &array_type = to_array_type(type);
```
don't you want a const here?

tautschnig replied:
`type` is now a `const typet &`, so all this is implicitly `const`.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+        result_type.subtype() = std::move(*renamed_subtype);
+
+      if(renamed_size != array_type.size())
+        result_type.size() = std::move(renamed_size);
```
the conditional doesn't seem really useful (but maybe it makes sense for the following commits)

tautschnig replied:
Yes, the later commit actually does `renamed_size.has_value()`.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
-    for(auto &component : components)
+    struct_union_typet::componentst::iterator comp_it =
+      s_u_type.components().begin();
+    for(auto &component : to_struct_union_type(type).components())
```
Why is there an iterator `comp_it` if we are using a range-for, this seems redundant (but maybe not relevent to this PR).

tautschnig replied:
No, this is not redundant: the iterator keeps track of which component needs to be updated (if any), while the ranged-for operates over the (`const`) source type.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+      auto renamed_subtype =
+        rename_type<level>(to_pointer_type(type).subtype(), irep_idt(), ns))
+    {
+      pointer_typet pointer_type = to_pointer_type(type);
```
Why copy the pointer type if we are going to replace its subtype? It seems better to construct a new one.

tautschnig replied:
Done, though I'm not sure this is strictly better (we need to parse the width).

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+  }
+  else if(type.id() == ID_symbol_type)
+  {
+    const symbolt &symbol = ns.lookup(to_symbol_type(type));
```
Only `symbol.type` is used, so declare a variable for that instead of `symbol`
:tiger:

tautschnig replied:
This specific bit is gone, but still applies below as you indicated.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -706,46 +756,100 @@ void goto_symex_statet::rename(
+  }
+  else if(type.id() == ID_struct_tag || type.id() == ID_union_tag)
+  {
+    const symbolt &symbol = ns.lookup(to_tag_type(type));
```
Same comment as :tiger: above

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -368,9 +368,15 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
+
+    if(renamed_object.has_value())
+    {
+      address_of_exprt result = address_of_expr;
```
Maybe construct a new object instead of copying address_of_expr since `.object()` is changed afterwards.

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
   {
-    expr=ssa_exprt(expr);
-    rename_address<level>(expr, ns);
+    ssa_exprt ssa(expr);
```
`ssa{expr}`

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
+        to_struct_union_type(renamed_struct_op->type());
+      const struct_union_typet::componentt &comp =
+        su_type.get_component(member_expr.get_component_name());
+      PRECONDITION(comp.is_not_nil());
```
I'm not sure this really qualifies as a precondition

tautschnig replied:
True, but also this was just what it was before (this code is now just indented one level less). Nevertheless I've changed into into a `DATA_INVARIANT`.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
+  }
+  else
+  {
+    exprt result = expr;
```
This copy is sometimes not used, but I don't know if there is a simple way around it.

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -359,11 +380,19 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
+        return {};
     }
+
+    ssa_exprt ssa(expr);
```
`ssa{expr}`

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -375,29 +404,47 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
   }
   else
   {
+    exprt result = expr;
```
this copy may be unused

tautschnig replied:
Done.

romainbrenguier commented:
@tautschnig https://github.com/diffblue/cbmc/pull/4286 has been merged, and this needs rebasing

tautschnig committed:
Remove requires_renaming

We return a renamed type if, and only if, renaming was necessary. Thus
we can do the is-it-necessary and the actual renaming in a single pass.

tautschnig committed:
Failing test to show that requires_renaming recurses

requires_renaming must not recurse into array subtypes in structs.

tautschnig commented:
@romainbrenguier @smowton All comments addressed.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -538,128 +611,170 @@ bool goto_symex_statet::l2_thread_write_encoding(
-    {
-      // this could go wrong, but we would have to re-typecheck ...
-      rename<level>(expr.type(), irep_idt(), ns);
+      return {};
```
The function is becoming quite big and it may be good to extract some static function for instance to make an optional `if_exprt` from an original expression and three optionals, but it doesn't need to be in this pull request.

romainbrenguier reviewed (approved):
Can you give some figures to show this reduces the number of `detach` calls, maybe just on an example?

smowton reviewed (approved):
OK for now, I'd like to look at using something like `expr_iterator` to do this without so much boilerplate though

romainbrenguier commented:
@smowton 
> OK for now, I'd like to look at using something like expr_iterator to do this without so much boilerplate though

I tried that but it doesn't really work, because of the way mutate is implemented we actually end-up doing more `detach` (even than the version without optional).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 71aec163b1f9dac403b12adb16066c1a6bf101cb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103873701

tautschnig commented:
I'm currently running benchmarks of develop vs. this branch to back up these changes by performance data. If those do not yield any negative results I'm inclined to merge this as-is and do the cleanup proposed by @romainbrenguier and @smowton in follow-up work.

tautschnig commented:
Here are my profiling results (on ReachSafety-ECA from SV-COMP):
develop performs 2959.21 symex steps per second, applying this PR we achieve 2844.20 steps per second (approximately 4% slowdown). The number of calls to `detach` reduces from 25,442,953,846 to 18,862,575,940. The amount of time spent in `rename<L2>(exprt, ...)` seems to increase from 434 seconds to 464 seconds, but it's not entirely clear what that time is spent on other than just doing recursive calls - it seems the recursive calls got a bit more expensive, which would seem fair given that we now have a return value to manipulate and some branching to do.

In my opinion this is no clear indication either way. I think the approach of not breaking sharing is cleaner, but it's certainly no performance win by itself.

@kroening @peterschrammel @romainbrenguier @smowton Can you please voice your opinions on whether we should do this or not?

romainbrenguier commented:
Since it doesn't have a positive impact on performances and makes the code more complicated, I would be in favor of not merging it.

romainbrenguier reviewed (changes requested):
I'm "requesting changes" so that this doesn't get merged by mistake.

smowton commented:
Agreed that the sharing breaking is a great thing, and it seems like it shouldn't be too difficult to get that without incurring a lot of extra cost tracking that. Could you find out if it reduces the lost time to return `exprt` rather than `optionalt<exprt>` and use `&old_expr.read() == &new_expr.read()` to determine if there has been a change? In particular this will enable return elision as we'll be directly returning an `exprt`.

tautschnig commented:
Closing as all this code has been substantially rewritten already and uses move semantics today.
